dsa/validate.go

```diff
@@ -1,51 +1,104 @@
 	"github.com/prebid/prebid-server/v2/openrtb_ext"
+	"github.com/prebid/prebid-server/v2/util/jsonutil"
+)
+
+// ObjectSignal represents publisher DSA object required statuses
+type ObjectSignal int8
+
+const (
+	ObjectSignalRequired               ObjectSignal = 2 // bid responses without DSA object will not be accepted
+	ObjectSignalRequiredOnlinePlatform ObjectSignal = 3 // bid responses without DSA object will not be accepted, Publisher is Online Platform
+)
+
+// PubSignal represents publisher rendering intentions
+type PubSignal int8
 
-	"github.com/buger/jsonparser"
+const (
+	PubSignalCannotRender PubSignal = 0 // publisher can't render
+	PubSignalCanRender    PubSignal = 1 // publisher could render depending on adrender
+	PubSignalWillRender   PubSignal = 2 // publisher will render
 )
 
+// BuyerSignal represents buyer/advertiser rendering intentions
+type BuyerSignal int8
+
 const (
-	// Required - bid responses without DSA object will not be accepted
-	Required = 2
-	// RequiredOnlinePlatform - bid responses without DSA object will not be accepted, Publisher is an Online Platform
-	RequiredOnlinePlatform = 3
+	BuyerSignalWontRender BuyerSignal = 0 // buyer/advertiser will not render
+	BuyerSignalWillRender BuyerSignal = 1 // buyer/advertiser will render
+)
+
+var (
+	ErrDsaMissing        = errors.New("DSA object missing when required")
+	ErrBehalfTooLong     = errors.New("DSA behalf exceeds limit of 100 chars")
+	ErrPaidTooLong       = errors.New("DSA paid exceeds limit of 100 chars")
+	ErrNeitherWillRender = errors.New("DSA publisher and buyer both signal will not render")
+	ErrBothWillRender    = errors.New("DSA publisher and buyer both signal will render")
 )
 
 // Validate determines whether a given bid is valid from a DSA perspective.
 // A bid is considered valid unless the bid request indicates that a DSA object is required
-// in bid responses and the object happens to be missing from the specified bid.
-func Validate(req *openrtb_ext.RequestWrapper, bid *entities.PbsOrtbBid) (valid bool) {
-	if !dsaRequired(req) {
-		return true
+// in bid responses and the object happens to be missing from the specified bid, or if the bid
+// DSA object contents are invalid
+func Validate(req *openrtb_ext.RequestWrapper, bid *entities.PbsOrtbBid) error {
+	reqDSA := getReqDSA(req)
+	bidDSA := getBidDSA(bid)
+
+	if dsaRequired(reqDSA) && bidDSA == nil {
+		return ErrDsaMissing
 	}
-	if bid == nil || bid.Bid == nil {
-		return false
+	if bidDSA == nil {
+		return nil
 	}
-	_, dataType, _, err := jsonparser.Get(bid.Bid.Ext, "dsa")
-	if dataType == jsonparser.Object && err == nil {
-		return true
-	} else if err != nil && err != jsonparser.KeyPathNotFoundError {
-		return true
+	if len(bidDSA.Behalf) > 100 {
+		return ErrBehalfTooLong
 	}
-	return false
+	if len(bidDSA.Paid) > 100 {
+		return ErrPaidTooLong
+	}
+	if reqDSA.PubRender == int8(PubSignalCannotRender) && bidDSA.AdRender != int8(BuyerSignalWillRender) {
```

Comment:
Is the cast necessary if they're all a int8 underlying type?

Reply:
Yes it is.
```
± |DSA-3438 → origin {45} ?:4 ✗| → go clean --testcache && go test -v ./dsa/
# github.com/prebid/prebid-server/v2/dsa [github.com/prebid/prebid-server/v2/dsa.test]
dsa/validate.go:67:25: invalid operation: reqDSA.PubRender == PubSignalWillRender (mismatched types int8 and PubSignal)
dsa/validate.go:67:67: invalid operation: bidDSA.AdRender == BuyerSignalWillRender (mismatched types int8 and BuyerSignal)
```